cluster/images/etcd/rollback/rollback.go

```diff
@@ -0,0 +1,318 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"log"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/coreos/etcd/etcdserver"
+	pb "github.com/coreos/etcd/etcdserver/etcdserverpb"
+	"github.com/coreos/etcd/etcdserver/membership"
+	"github.com/coreos/etcd/mvcc/backend"
+	"github.com/coreos/etcd/mvcc/mvccpb"
+	"github.com/coreos/etcd/pkg/pbutil"
+	"github.com/coreos/etcd/pkg/types"
+	"github.com/coreos/etcd/raft/raftpb"
+	"github.com/coreos/etcd/snap"
+	"github.com/coreos/etcd/store"
+	"github.com/coreos/etcd/wal"
+	"github.com/coreos/etcd/wal/walpb"
+)
+
+var (
+	migrateDatadir string
+	ttl            time.Duration
+)
+
+func init() {
+	flag.StringVar(&migrateDatadir, "data-dir", "", "Path to the data directory")
+	flag.DurationVar(&ttl, "ttl", time.Hour, "TTL of event keys (default 1 hour)")
+	flag.Parse()
+}
+
+func main() {
+	dbpath := path.Join(migrateDatadir, "member", "snap", "db")
+
+	be := backend.New(dbpath, time.Second, 10000)
+	tx := be.BatchTx()
+
+	st := store.New(etcdserver.StoreClusterPrefix, etcdserver.StoreKeysPrefix)
+	expireTime := time.Now().Add(ttl)
+
+	tx.Lock()
+	err := tx.UnsafeForEach([]byte("key"), func(k, v []byte) error {
+		kv := &mvccpb.KeyValue{}
+		kv.Unmarshal(v)
+
+		// This is compact key.
+		if !strings.HasPrefix(string(kv.Key), "/") {
+			return nil
+		}
+
+		ttlOpt := store.TTLOptionSet{}
+		if kv.Lease != 0 {
+			ttlOpt = store.TTLOptionSet{ExpireTime: expireTime}
+		}
+
+		if !isTombstone(k) {
+			_, err := st.Set(path.Join("1", string(kv.Key)), false, string(kv.Value), ttlOpt)
+			if err != nil {
+				return err
+			}
+		} else {
+			st.Delete(string(kv.Key), false, false)
+		}
+
+		return nil
+	})
+	if err != nil {
+		panic(err)
+	}
+	tx.Unlock()
+
+	traverseAndDeleteEmptyDir(st, "/")
+
+	metadata, hardstate, oldSt := rebuild(migrateDatadir)
+
+	// In the following, it's low level logic that saves metadata and data into v2 snapshot.
+
+	if err := os.RemoveAll(migrateDatadir); err != nil {
+		panic(err)
+	}
+	if err := os.MkdirAll(path.Join(migrateDatadir, "member", "snap"), 0700); err != nil {
+		panic(err)
+	}
+	walDir := path.Join(migrateDatadir, "member", "wal")
+
+	w, err := wal.Create(walDir, metadata)
+	if err != nil {
+		panic(err)
+	}
+	err = w.SaveSnapshot(walpb.Snapshot{Index: hardstate.Commit, Term: hardstate.Term})
+	if err != nil {
+		panic(err)
+	}
+	w.Close()
+
+	nodeIDs := []uint64{}
+	event, err := oldSt.Get(etcdserver.StoreClusterPrefix, true, false)
+	if err != nil {
+		panic(err)
+	}
+	// searching all metadata nodes and:
+	// - update store
+	// - update Nodes for ConfState
```

Comment:
Sorry - I went through the code below and I still don't understand what it is doing. Can you please clarify the comment (including information why this is needed?)